ecp5/arch_place.cc

```diff
@@ -190,20 +190,273 @@ bool Arch::isBelLocationValid(BelId bel, bool explain_invalid) const
+                              ci->type.c_str(getCtx()),
+                              ci->name.c_str(getCtx()),
+                              port.c_str(getCtx()));
+                }
+                assigned_nets[j] = net;
+                remap_ports[i] = ports[j];
+                log_info("DSP: %s '%s': Connection to %s remapped to %s\n",
+                         ci->type.c_str(getCtx()), ci->name.c_str(getCtx()),
+                         ports[i].c_str(getCtx()), ports[j].c_str(getCtx()));
+            }
+        } else if (*assigned != assigned_nets[i]) {
+            // If the net has been assigned already and to a different
+            // port than this one, we'll remap the port and attributes
+            // to point to the already-assigned port.
+            size_t j = std::distance(assigned_nets.cbegin(), assigned);
+            remap_ports[i] = ports[j];
+            log_info("DSP: %s '%s': Connection to %s remapped to %s\n",
+                     ci->type.c_str(getCtx()), ci->name.c_str(getCtx()),
+                     ports[i].c_str(getCtx()), ports[j].c_str(getCtx()));
+        }
+    }
+
+    // Second, connect each port to its assigned net.
+    for (size_t i = 0; i < ports.size(); i++) {
+        IdString port = ports[i];
+        ci->disconnectPort(port);
+        if (assigned_nets[i] != nullptr) {
+            ci->connectPort(port, assigned_nets[i]);
+        }
+    }
+
+    // Third, remap any parameters that refer to old ports to refer to the
+    // new port instead.
+    for (auto remap_param : remap_params) {
+        auto param = ci->params.find(remap_param);
+        if (param == ci->params.end())
+            continue;
+        for (size_t i = 0; i < remap_ports.size(); i++) {
+            Property &prop = param->second;
+            if (remap_ports[i] != IdString()
+                && prop.is_string
+                && prop.str == ports[i].str(getCtx())
+            ) {
+                prop = Property(remap_ports[i].str(getCtx()));
+                break;
+            }
+        }
+    }
+
+    // Finally, only when remapping CLK ports, also move any `CLKn_DIV`
+    // to the new clock port.
+    const std::array<IdString, 4> clk_div_params = {
+        id_CLK0_DIV, id_CLK1_DIV, id_CLK2_DIV, id_CLK3_DIV};
+    std::array<Property, 4> new_clk_div_props = {};
+    if (ports[0] == id_CLK0) {
+        for (size_t i = 0; i < 4; i++) {
+            if (remap_ports[i] == IdString())
+                continue;
+            auto param = ci->params.find(clk_div_params[i]);
+            if (param == ci->params.end())
+                continue;
+            size_t j = std::distance(
+                ports.cbegin(),
+                std::find(ports.cbegin(), ports.cend(), remap_ports[i]));
+            if (j != ports.size()) {
+                new_clk_div_props[j] = param->second;
+            }
+        }
+
+        for (size_t i = 0; i < 4; i++) {
+            if (new_clk_div_props[i].is_string) {
```

Comment:
would just like to double check what this `is_string` is about? I think it's just checking if `new_clk_div_props[i]` has been set (it will be false if it hasn't), if I understand this correctly maybe `if (new_clk_div_props[i] == Property())` would be cleaner?

Reply:
Yep, it's just checking whether `new_clk_div_props[i]` was set in the previous loop. I think I just had `is_string` in my head from the previous chunk where I do actually access properties as strings. The only valid settings for this property are `"ENABLED"` and `"DISABLED"` so I guess it would always be a string, but nevertheless your suggestion is neater.

I think you mean `if (new_clk_div_props[i] != Property())` (`!=` instead of `==`)?

Reply:
Thanks for confirming - and yes, I did mean `!=`...

Reply:
I've updated the PR with this change, thanks!